Src/zipkin4net/Src/Internal/Recorder/Recorder.cs

```diff
@@ -0,0 +1,250 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using zipkin4net.Tracers.Zipkin;
+using zipkin4net.Tracers.Zipkin.Thrift;
+using zipkin4net.Utils;
+using Span = zipkin4net.Tracers.Zipkin.Span;
+
+namespace zipkin4net.Internal.Recorder
+{
+    internal class Recorder : IRecorder
+    {
+        private readonly IEndPoint _localEndPoint;
+        private readonly MutableSpanMap _spanMap;
+
+        internal Recorder(IEndPoint localEndPoint, IReporter reporter)
+        {
+            _localEndPoint = localEndPoint;
+            _spanMap = new MutableSpanMap(reporter, new Statistics()); //todo inject statistics
+        }
+
+        public void Start(ITraceContext context)
+        {
+            Start(context, TimeUtils.UtcNow);
+        }
+
+        public void Start(ITraceContext context, DateTime timestamp)
+        {
+            _spanMap.GetOrCreate(context, (t) => new Span(t, timestamp));
+        }
+
+        public void Name(ITraceContext context, string name)
+        {
+            var span = GetSpan(context);
+            span.Name = name;
+        }
+
+        [Obsolete("Backward compatibility. Exclusively for ZipkinAnnotationVisitor")]
+        internal void ServiceName(ITraceContext context, string serviceName)
+        {
+            var span = GetSpan(context);
+            span.ServiceName = serviceName;
+        }
+
+        [Obsolete("Backward compatibility. Exclusively for ZipkinAnnotationVisitor")]
+        internal void EndPoint(ITraceContext context, IPEndPoint endPoint)
+        {
+            var span = GetSpan(context);
+            span.Endpoint = endPoint;
+        }
+
+        private Span GetSpan(ITraceContext context)
+        {
+            return _spanMap.Get(context);
+        }
+
+        public void Kind(ITraceContext context, SpanKind kind)
+        {
+            var span = GetSpan(context);
+            span.SpanKind = kind;
+        }
+
+        public void RemoteEndPoint(ITraceContext context, IEndPoint remoteEndPoint)
+        {
+            var span = GetSpan(context);
+            if (span.SpanKind.HasValue)
```

Comment:
you could consider doing this late, in case there's a race on setting kind, ex right before marshaling